src/Lecture4.hs

```diff
@@ -223,7 +287,18 @@ implement the next task.
 -}
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
+combineRows (a :| as) = go (rowToStats a) as
+    where go :: Stats -> [Row] -> Stats
+          go !b [] = b
+          go !b (c:cs) = go (d (rowToStats c <> b)) cs
+          d x = trace ("a: " ++ show x) x
```

Comment:
Did this:
```haskell
instance Semigroup Stats where
    (<>) a b = Stats
                { statsTotalPositions = statsTotalPositions a <> statsTotalPositions b
                , statsTotalSum = statsTotalSum a <> statsTotalSum b
                , statsAbsoluteMax = statsAbsoluteMax a <> statsAbsoluteMax b
                , statsAbsoluteMin = statsAbsoluteMin a <> statsAbsoluteMin b
                , statsSellMax = apl (statsSellMax a <> statsSellMax b)
                , statsSellMin = apl (statsSellMin a <> statsSellMin b)
                , statsBuyMax = apl (statsBuyMax a <> statsBuyMax b)
                , statsBuyMin = apl (statsBuyMin a <> statsBuyMin b)
                , statsLongest = statsLongest a <> statsLongest b
                }
                where apl Nothing = Nothing
                      apl (Just !something) = Just something
```
But not sure whether it's "smarter" :) Looks hackish.

Reply:
There exist the Haskell package `strict` that implements a stricter version of `Maybe`, so you can use it instead of standard `Maybe`

* https://hackage.haskell.org/package/strict-0.4.0.1/docs/Data-Strict-Maybe.html

I wouldn't call it a "hack" if it fixes fallacies of the standard library 🙂